File: code/bit_manipulation/src/addition_using_bits/addition_using_bits.py
# Part of Cosmos by OpenGenus Foundation

# Function adds 2 numbers a and b without using + operator
def addition(a, b):
    while b != 0:
        c = a & b  # carry contains set bits of both a and b
        a = a ^ b  # a ^ b is equal to 1 if, the bits of a and b are different
        b = c << 1  # carry is right-shifted by 1
    return a


# Driver code
if __name__ == "__main__":
    print(addition(5, 9))
    print(addition(4, 3))

File: code/bit_manipulation/src/addition_using_bits/addition_using_bits.go
package main

import "fmt"

// Part of Cosmos by OpenGenus Foundation
func bitwiseAddition(n, m int) int {
  for m != 0 {
    var carry int = n & m; // Variable carry keeps track of bits that carry over
    n = n ^ m;             // This adds up the individual bits
    m = carry << 1;        // Shift carry over bits so they can be added
  }
  return n;
}

func bitwiseAdditionRecursive(n, m int) int {
  if m == 0 {
    return n
  } else {
    var carry int = n & m
    return bitwiseAdditionRecursive(n ^ m, carry << 1)
  }
}
